torchtune/datasets/rfc_iterable_dataset.md

```diff
@@ -0,0 +1,439 @@
+    def __init__(
+        self,
+        *,
+        load_args: Dict,
+        message_transform: Callable,
+        tokenizer: Callable,
+        recipe_transform: Callable,
+        shuffle_buffer_size: Optional[int] = 1000,
+        seed: Optional[int] = 42,
+        num_shards_per_worker: int = 16,
+        weight: float = 1.0,
+        filter_args: Optional[Dict] = None,
+        *args, **kwargs
+    ):
+        """Initialize a single dataset with its specific transformations."""
+        self.weight = weight
+
+        world_size = 1
+        if torch.distributed.is_initialized():
+            world_size = torch.distributed.get_world_size()
+
+        # TODO: Maybe num shards should be based on dataset size, if we know it
+        num_shards = world_size * num_shards_per_worker
+        ds = load_dataset(**load_args)
+        ds = ds.to_iterable_dataset(num_shards)
+
+        if filter_args:
+            function = filter_args.get("function", None)
+            if function and not isinstance(function, Callable):
+                raise ValueError(
+                    f"filter_args['function'] must be a callable. Found {type(function)}"
+                )
+            # https://huggingface.co/docs/datasets/v3.6.0/en/stream#filter
+            ds = ds.filter(**filter_args)
+
+        def _apply_transforms(sample):
+            sample = message_transform(sample)
+            sample = tokenizer(sample)
+            return recipe_transform(sample)
+
+        ds = ds.map(_apply_transforms)  # lazy
+
+        if shuffle_buffer_size and shuffle_buffer_size > 0:
+            ds = ds.shuffle(shuffle_buffer_size, seed)
+
+        # Distribute
+        if world_size > 1:
+            ds = split_dataset_by_node(
+                ds,
+                rank=torch.distributed.get_rank(),
+                world_size=world_size,
+            )
+
+        self.ds = ds
+
+    def __iter__(self):
+        # Expose the for loop so extra logic can be added here, e.g. drop if no trainable tokens
+        # TODO: should we add try/except to handle/log errors?
+        for sample in self.ds:
+            yield sample
+
+    def state_dict(self):
+        state_dict = self.ds.state_dict()
+        state_dict["weight"] = self.weight
+        return state_dict
+
+    def load_state_dict(self, state_dict):
+        self.weight = state_dict.pop("weight")
+        self.ds.load_state_dict(state_dict)
+```
+
+#### Setup Data
+Method in recipes/full_distributed.py or *utility* used in the recipe
+
+```python
+from datasets import interleave_datasets, split_dataset_by_node
+from torchtune.models.tokenizers import ModelTokenizer
+import torch
+
+# NOTE: Mixed feelings about passing multiple ConfigDict to setup_data.
+# Hard for users to know what they should contain.
+# On the other hand:
+# i) setup_data doesn't need to make assumptions about the configs
+# ii) we already do it currently
+# Alternative: use dataclasses?
+
+def setup_data(
+    dataset_cfg: ConfigDict,
```

Comment:
This is at the core of why I wouldn't get into the business of constructing HF datasets ourselves.

I'd much rather standarize Tune's interface around generic iterable datasets, and this is an interface that very naturally plugs into HF datasets. If and when a user wants to use one of those, they can construct it by themselves and simply pass it over

Reply:
In general, I agree that eventually Tune's interface should be based on generic iterable principles, but in this case I think it's a little of "don't pre-optimize."

The small `datasets` library gives us a lot of power right now and I don't think we have enough signal to know really what a perfect generic iterable design would be.